### ISSUE TITLE
CreateKnativeService does not sync to desired state in non-happy path cases

### ISSUE THREAD

**Issue by rakelkar** [kubeflow/kfserving#41]
@yuzisun The current implementation of CreateKnativeService will not sync cluster state to spec desired state in the following cases (admittedly off the happy path):
1. User modifies the default spec while a canary is active. Default revision will remain unchanged
2. New cluster (does not have any existing revisions) syncs serving CRD with default and canary. Will only provision canary?

I think these issues stem from the fact that this logic depends on cluster state (revision state). Can we do some other knative thing to resolve this? Happy to try to fix if can point me in the right kn direction!

https://github.com/kubeflow/kfserving/blob/master/pkg/reconciler/ksvc/resources/knative_service.go
```go
	if kfsvc.Spec.Canary == nil || kfsvc.Spec.Canary.TrafficPercent == 0 {
		//TODO(@yuzisun) should we add model name to the spec, can be different than service name?
		container = CreateModelServingContainer(kfsvc.Name, &kfsvc.Spec.Default)
		revisions = []string{knservingv1alpha1.ReleaseLatestRevisionKeyword}
		routingPercent = 0
	} else {
		container = CreateModelServingContainer(kfsvc.Name, &kfsvc.Spec.Canary.ModelSpec)
		revisions = []string{kfsvc.Status.Default.Name, knservingv1alpha1.ReleaseLatestRevisionKeyword}
		routingPercent = kfsvc.Spec.Canary.TrafficPercent
	}
```

**Comment by rakelkar** [comment]
Maybe use route instead? Related KNative issue: https://github.com/knative/serving/issues/2721

From the route spec: https://github.com/knative/serving/blob/master/pkg/apis/serving/v1beta1/route_types.go
```
// Route is responsible for configuring ingress over a collection of Revisions.
// Some of the Revisions a Route distributes traffic over may be specified by
// referencing the Configuration responsible for creating them; in these cases
// the Route is additionally responsible for monitoring the Configuration for
// "latest ready revision" changes, and smoothly rolling out latest revisions.
```

**Comment by yuzisun** [comment]
> @yuzisun The current implementation of CreateKnativeService will not sync cluster state to spec desired state in the following cases (admittedly off the happy path):
> 
> 1. User modifies the default spec while a canary is active. Default revision will remain unchanged

Thanks for bringing this up! I have thought about this and like to add the validation against invalid states before starting the reconciliation in controller, there are a few other cases like adding canary while default is not ready yet. In order to do this we need to start watching the knative revision resources, will be a bit change.

**Comment by yuzisun** [comment]
> @yuzisun The current implementation of CreateKnativeService will not sync cluster state to spec desired state in the following cases (admittedly off the happy path):
> 
> 1. User modifies the default spec while a canary is active. Default revision will remain unchanged
> 2. New cluster (does not have any existing revisions) syncs serving CRD with default and canary. Will only provision canary?
You are right this also depends on revision states, we will need to check whether the default revision exists in the cluster or not.

**Comment by yuzisun** [comment]
> Maybe use route instead? Related KNative issue: [knative/serving#2721](https://github.com/knative/serving/issues/2721)
> 
> From the route spec: https://github.com/knative/serving/blob/master/pkg/apis/serving/v1beta1/route_types.go
> 
> ```
> // Route is responsible for configuring ingress over a collection of Revisions.
> // Some of the Revisions a Route distributes traffic over may be specified by
> // referencing the Configuration responsible for creating them; in these cases
> // the Route is additionally responsible for monitoring the Configuration for
> // "latest ready revision" changes, and smoothly rolling out latest revisions.
> ```

I think what Matt on that ticket saying is that it supports automatic canary rollout at route level but not service, when you rollout a new revision the route will not assign the traffic to the new revision until it is in ready state(passing readiness probe check). For service it is a bit more than that since `Ready` state is defined by some metric, e.g in our case whether the model is good or not.

**Comment by cliveseldon** [comment]
Do we need to check these invalid conditions against proper "gitops" usage? For instance, updating the resource with a canary before the original "default" is ready seems fine to me by the rules of "gitops" and "declarative state" ideas - the controller should simply update to the new default+canary scenario gracefully. Or am I missing something here why that is not valid?

**Comment by rakelkar** [comment]
@cliveseldon, Agree - the code should sync cluster state to desired state specified in the spec (ie for both default and canary) without any additional rules or policies. The issue is that the current implementation do that. The implementation assumes that a canary will only be set after default has been set. So if a canary in the is set the spec, it happily creates a new revision for canary assuming that  current serving revision must correspond to default. The code doesnt validate this assumption, and the cases I stated above are two examples of where it is invalid.

I cant figure out any way to resolve this easily while relying on revisions to map to default and canary.

For instance consider:
>cluster state: default: a1, canary: b1
oops discovered an issue with default that needs to be patched asap
>desired state: default a1*, canary: b1
current impl will not update default (and cant - since it cant update a revision).

Also the current impl does not appear to deploy canary if trafficPercent=0...?
Also looking at the code again, shouldnt routingPercent be set to 100 for the default case (when kfsvc.Spec.Canary == nil)...

Am I missing something really basic in my interpretation of the code?

**Comment by cliveseldon** [comment]
I think you are right there needs to be some discussion of whether just using Knative services can achieve the functionality we desire. I assumed the code was a starting point.

One thing I thought about was the control plane doc discusses switching canary traffic to be default as final step in canary rollout. For this you would need to keep track of how revisions match the current desired state and handle either creation of a new revision or utilization of existing revision that matches spec. I'm not clear how that is to be implemented.

**Comment by yuzisun** [comment]
@cliveseldon @rakelkar I agree ideally we shouldn't need to check these conditions, but syncing both default and canary worries me a bit since that could result in multiple updates to knative service in a single reconcile step(potential race condition). Also in the above case when default is bad, I feel user should really just delete canary and fix default, it is a more clean workflow, with the current implementation it ignores the default spec change until you delete canary.

**Comment by yuzisun** [comment]
> Also the current impl does not appear to deploy canary if trafficPercent=0...?

`trafficPercent = 0` means routing to default and once user set `trafficPercent > 0` canary gets deployed, I am not seeing deploying canary is needed when `trafficPercent = 0`. Looks our disconnect is that you are always expecting syncing the entire spec but i am trying to do single update at a time :).
> Also looking at the code again, shouldnt routingPercent be set to 100 for the default case (when kfsvc.Spec.Canary == nil)...

in knative routingPercent 100 is not allowed, when there is only one revision on the list that means 100 percent traffic is going there and `routingPercent` field does not take effect.
> 
> Am I missing something really basic in my interpretation of the code?

**Comment by ellis-bigelow** [comment]
Just catching up to this thread. I think it's possible to just model this with two Knative Revisions and one Route that balances traffic.

**Comment by yuzisun** [comment]
@ellis-bigelow @rakelkar @cliveseldon, After discussing with Knative team and Ellis is right just that revisions are immutable and we need to use configuration, I am pretty sold that we can go for two configurations and one route that splits the traffic, trying out that now.

**Comment by ellis-bigelow** [comment]
Cool!

**Comment by cliveseldon** [comment]
That's great. 

  * Why are 2 configurations needed rather than 1 with revisions for default and canary?
  * Will this allow gitops where just a route change is made when you move canary to default? Just want to check we can provide the required functionality without creating new revisions in these circumstances.

**Comment by rakelkar** [comment]
closed by #51
